Unexpected TensorBoard behavior

seungjaeryanlee opened the pull request:
Thank you for the great feedback for my last blocker (#1). I moved the RND network to the agent so that it can be trained. To do so, I created a `RndDqnAgent` class that inherits the `DqnAgent` class. The `_loss()` function has been modified to compute RND loss as well, which is used in `_train()` function as intrinsic reward and to train the RND network.

However, I am noticing a strange behavior for TensorBoard. 

![Screenshot from 2019-06-28 18-09-29](https://user-images.githubusercontent.com/6107926/60331718-44524380-99d0-11e9-8388-e836782628d1.png)

1. The value is only updated every 1k steps, and
2. RND loss is not being plotted even though it `summary.scalar()` is invoked.

Do you have any ideas about what could be the cause of the problem?

(Please ignore the changes in `rnd_wrapper.py` and `nest_utils.py`, since they will be reverted.)

seungjaeryanlee committed:
Add stub RNDWrapper

seungjaeryanlee committed:
Merge branch 'master' into feat/rnd-core

seungjaeryanlee committed:
Create RNDDQN training script

Uses stub rnd_wrapper

seungjaeryanlee committed:
Fix bug where new timestep was not assigned

seungjaeryanlee committed:
Set correct dtype for stub intrinsic rewards

seungjaeryanlee committed:
Implement RNDNetwork

seungjaeryanlee committed:
Debug spec_shape error

seungjaeryanlee committed:
Merge branch 'master' into feat/rnd-core

seungjaeryanlee committed:
Add print statements for debugging

seungjaeryanlee committed:
Merge branch 'master' into feat/rnd-core

seungjaeryanlee committed:
Use BoundedTensorSpec for observation and action specs

seungjaeryanlee committed:
Merge branch 'master' into feat/rnd-core

seungjaeryanlee committed:
Remove manually conversion to tensor

seungjaeryanlee committed:
Remove duplicate validate_specs() method

seungjaeryanlee committed:
Merge branch 'master' into feat/rnd-core

seungjaeryanlee committed:
Update rnd_network.py docstring

seungjaeryanlee committed:
Remove resolved TODO statement

Network state is used for recurrent layers

seungjaeryanlee committed:
Remove RNDNetwork

Can be replaced with EncodingNetwork for now

seungjaeryanlee committed:
Replace v1 train_eval script with v2

seungjaeryanlee committed:
Compute MSE of target and predictor network

seungjaeryanlee committed:
Merge branch 'master' into feat/rnd-core

seungjaeryanlee committed:
Merge branch 'master' into feat/rnd-core

seungjaeryanlee committed:
Add stub rnddqn_agent.py

seungjaeryanlee committed:
Add RND parameters to RndDqnAgent

seungjaeryanlee committed:
Add buggy training code for RndDqnAgent

seungjaeryanlee committed:
Stop using RNDWrapper from train_eval_rnd

seungjaeryanlee committed:
Debug RndDqnAgent

seungjaeryanlee committed:
Use squared loss as default RND loss function

seungjaeryanlee committed:
Remove redundancy of RndDqnAgent by inheriting DqnAgent

seungjaeryanlee committed:
Remove unused parameters for RndDqnAgent._rnd_loss()

seungjaeryanlee committed:
Compute and backprop RND loss with TD loss

seungjaeryanlee committed:
Use RND loss as intrinsic reward

oars commented on tf_agents/agents/dqn/rnddqn_agent.py:
```diff
@@ -0,0 +1,456 @@
+
+      with tf.name_scope('Losses/'):
+        tf.compat.v2.summary.scalar(
+            name='loss', data=avg_rnd_loss, step=self.train_step_counter)
```
```suggestion
            name='rnd_loss', data=avg_rnd_loss, step=self.train_step_counter)
```

seungjaeryanlee replied:
![Screenshot from 2019-06-30 01-13-01](https://user-images.githubusercontent.com/6107926/60386748-4b627a00-9ad4-11e9-8f99-b95c15b93d94.png)

This was indeed the problem! Thank you Oscar

oars commented on tf_agents/agents/dqn/rnddqn_agent.py:
```diff
@@ -0,0 +1,456 @@
+      rnd_target, _ = self._target_rnd_network(time_steps.observation,
+                                               time_steps.step_type)
+
+      rnd_losses = element_wise_squared_loss(rnd_prediction, rnd_target)
```
```suggestion
      rnd_losses = rnd_loss_fn(rnd_prediction, rnd_target)
```

seungjaeryanlee replied:
Thank you for noticing it!

oars commented on tf_agents/agents/dqn/rnddqn_agent.py:
```diff
@@ -0,0 +1,456 @@
+        td_targets = compute_td_targets(
+            next_q_values,
+            # TODO Should reward_scale_factor multiply both intrinsic and extrinsic rewards?
+            rewards=reward_scale_factor * (next_time_steps.reward + avg_rnd_loss),
```
```suggestion
            rewards=reward_scale_factor * (next_time_steps.reward + rnd_losses),
```

seungjaeryanlee replied:
Made the change!

Also, I found out that I was using `element_wise_squared_loss()` for RND, which made `rnd_losses.shape == (64, 100)` and `avg_rnd_loss.shape == ()` (64 is the experience batch size, 100 is the final fc layer size). 

I created a new function `mean_squared_loss()`, so now `rnd_losses.shape == (64,)` and `avg_rnd_loss.shape == ()`.

ebrevdo committed:
Add tfa.networks.SequentialLayer, a keras Layer alternative to keras Sequence model.

This one should be more friendly with Networks, tf.function, and SavedModel..

PiperOrigin-RevId: 255677987
Change-Id: Iab138fa97a54e63b39668c2cf7a4c47bf40a6309

seungjaeryanlee committed:
Merge branch 'master' into feat/rnd-core

seungjaeryanlee committed:
Fix duplicate Tensorboard name

seungjaeryanlee committed:
Use rnd_loss_fn given by parameter

seungjaeryanlee committed:
Use correct MSE loss function for RND

seungjaeryanlee committed:
Update variable names and docstrings

seungjaeryanlee committed:
Use individual RND loss for each experience

seungjaeryanlee commented:
Closing this PR since all feedback has been incorporated and all questions have been resolved. Thank you so much!
